text/3720-labeled-match.md

```diff
@@ -0,0 +1,1452 @@
+- Feature Name: `loop_match`
+- Start Date: 2024-09-26
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/3720)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC adds `loop match`:
+
+- a `loop` and `match` can be fused into a `loop match <scrutinee> { /* ... */ }`
+- a `loop match` can be targeted by a `continue <value>`. The `value` is treated as a replacement operand to the `match` expression.
+
+The state transitions (going from one branch of the match to another) can be annotated with the `const` keyword, providing more accurate CFG information to the backend. That means:
+
+- more valid programs are accepted, increasing expressivity
+- the backend can generate better code, leading to significant speedups
+
+### Basic example
+
+A `loop match` is similar to a `match` inside of a `loop`, with a mutable variable being updated to move to the next state. For instance, these two functions are semantically equivalent:
+
+```rust
+fn loop_match() -> Option<u8> {
+    loop match 1u8 {
+        1 => continue 2,
+        2 => continue 3,
+        3 => break Some(42),
+        _ => None
```

Comment:
The `loop match` syntax is advertised as *a `loop` and `match` can be combined into a `loop match`*, but the proposed semantics is clearly not the case:

```rust
// This will loop forever.
let a = loop { match true {
    true => {},
    false => unreachable!(),
}};

// This will assign `()` to `b`
let b = loop match true {
    true => {}, // a non-diverging value `x` implicitly becomes `break x;`
    false => unreachable!(),
};
```

For clarity it is better to force every block inside the `loop match` to be diverging (have type `!`), similar to the `else` block in a `let else` declaration. Yes this makes it more tedious to write but it will clear much confusion to read.